**Universidad Internacional de La Rioja (UNIR) - Máster Universitario en Inteligencia Artificial - Procesamiento del Lenguaje Natural** 

***
Datos del alumno (Nombre y Apellidos): **Jefferson Rodríguez Pallares**

Fecha:20/04/2026
***

<span style="font-size: 20pt; font-weight: bold; color: #0098cd;">Trabajo: Caracterización de textos</span>

**Objetivos** 

Con esta actividad se tratará de que el alumno se familiarice con el manejo de la librería spacy, así como con los conceptos básicos de manejo de las técnicas NER

**Descripción**

En esta actividad debes procesar de forma automática un texto en lenguaje natural para detectar características básicas en el mismo, y para identificar y etiquetar las ocurrencias de conceptos como localización, moneda, empresas, etc.

En la primera parte del ejercicio se proporciona un código fuente a través del cual se lee un archivo de texto y se realiza un preprocesado del mismo. En esta parte el alumno tan sólo debe ejecutar y entender el código proporcionado.

En la segunda parte del ejercicio se plantean una serie de preguntas que deben ser respondidas por el alumno. Cada pregunta deberá responderse con un fragmento de código fuente que esté acompañado de la explicación correspondiente. Para elaborar el código solicitado, el alumno deberá visitar la documentación de la librería spacy, cuyos enlaces se proporcionarán donde corresponda.

# Parte 1: carga y preprocesamiento del texto a analizar

Observa las diferentes librerías que se están importando.

In [1]:
!python -m spacy download es_core_news_sm

     ---------------------------------------- 0.0/12.9 MB ? eta -:--:--
     ------------------------------- ------- 10.5/12.9 MB 93.8 MB/s eta 0:00:01
     ---------------------------------------- 12.9/12.9 MB 50.5 MB/s  0:00:00
[+] Download and installation successful
You can now load the package via spacy.load('es_core_news_sm')


In [2]:
import pathlib
import spacy
import pandas as pd
from spacy import displacy
import csv
import es_core_news_md

El siguiente código simplemente carga y preprocesa el texto. Para ello, lo primero que hace es cargar un modelo de lenguaje previamente entrenado. En este caso, se utiliza <i>es_core_news_md</i>: 

https://spacy.io/models/es#es_core_news_md


In [3]:
nlp = es_core_news_md.load()

El objeto <i>nlp</i> permite utilizar el modelo de lenguaje cargado, de forma que se puede procesar un texto y obtenerlo en su versión preprocesada. Así, nos permite realizar las diferentes tareas. En este caso, vamos a utilizar el pipeline para hacer un preprocesamiento básico, que consiste en tokenizar el texto.

In [4]:
filename = "02Dataset_anonimizado.csv"

data = pd.read_csv(filename, delimiter=';', encoding='latin-1')

C:\Users\jefferson.rodriguez\AppData\Local\Temp\ipykernel_42084\2655521151.py:3: DtypeWarning: Columns (0: TIPO DE ODIO, 1: TONO HUMORISTICO, 2: MODIFICADOR) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv(filename, delimiter=';', encoding='latin-1')


El código anterior carga el archivo CSV (opcionalmente con un límite de líneas a leer) y genera la variable <i>data</i>, que contiene un Dataframe (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) con los datos leídos del CSV.

Te vendrá bien conocer la siguiente documentación:
<ul>
    <li>https://spacy.io/api/doc</li>
    <li>https://spacy.io/api/token</li>
    <li>https://spacy.io/api/morphology#morphanalysis</li>
</ul>

### Playground

Utiliza este espacio para hacer pruebas y ensayos con las variables generadas con el código previo. A modo de ejemplo, se ofrece código que realiza las siguientes tareas: 


- leer un número dado de líneas del Dataframe y generar dos listas con los valores (se pueden leer directamente del DataFrame, se muestra el ejemplo como una opción más)
- procesar el texto de cada comentario


Para procesarlo, hay utilizar el objeto <i>nlp</i> y así obtener objetos de la clase <i>Doc</i> (https://spacy.io/api/doc)

Visita la documentación de dicha clase y experimenta probando las diferentes funciones y atributos 

In [8]:
# Puedes insertar aquí código de pruebas para experimentar con las diferentes funciones y atributos de 'doc'.
#print(data["CONTENIDO A ANALIZAR"][1])
#print(data["INTENSIDAD"][1])
doc = []
value = []

#con el bucle, generamos sendas listas con los comentarios ya parseados y con el valor de intensidad
for i in range(0, 13):
    
    #en un primer paso se parsea el comentario. En el segundo paso se añade el objeto a la lista
    tmp_doc = nlp(data["CONTENIDO A ANALIZAR"][i])
    doc.append(tmp_doc)
    
    #en un primer paso extrae el valor. En el segundo paso se añade el valor a la lista
    tmp_value = data["INTENSIDAD"][i]
    value.append(tmp_value)


#ejemplo de cómo recorrer un comentario palabra por palabra    
for token in doc[1]:
    print(token)

el
real
madrid
ha
puesto
punto
y
final
a
su
andadura
en
la
copa
del
rey
en
el
primer
escalón
.
los
de
zidane
han
caí­do
ante
el
alcoyano
,
de
segunda
b
,
a
pesar
de
empezar
ganando
y
jugar
con
un
hombre
menos
en
la
prórroga
.
el
técnico
francés
dispuso
un
equipo
plagado
de
los
menos
habituales
,
con
vinicius
y
mariano
en
ataque
.
ninguno
de
los
dos
logró
crear
ocasiones
.
fue
militao
el
que
marcó
el
gol
del
madrid
,
justo
antes
del
descanso
.
en
la
segunda
parte
intentaron
cerrar
el
partido
,
pero
sin
el
colmillo
suficiente
y
el
modesto
alcoyano
aprovechó
un
córner
para
empatar
el
partido
a
cinco
minutos
para
el
final
.
el
empate
sentó
como
un
jarro
de
agua
frí­a
a
los
blancos
,
que
lo
intentaron
en
el
tiempo
extra
a
falta
de
cinco
minutos
,
el
casanova
consiguió
el
gol
más
importante
de
su
vida
,
que
vale
la
clasificación
para
octavos
de
la
copa
.
el
madrid
de
zidane
queda
apeado
del
torneo
una
vez
más
,
por
lo
que
el
francés
se
quedará
sin
pelear
por
el
único
tí­tulo
que
no
ha
conseg

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 1.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántos registros contiene el corpus?</span>

In [5]:
# Revisar la estructura del df recien cargado
print(data.info(),"\n","\n");
#Revisar si existen nulos
print("Los datos nulos por columna:\n", data.isnull().sum());



<class 'pandas.DataFrame'>
RangeIndex: 575029 entries, 0 to 575028
Data columns (total 16 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   MEDIO                 575029 non-null  str    
 1   SOPORTE               574896 non-null  str    
 2   URL                   574773 non-null  str    
 3   TIPO DE MENSAJE       574767 non-null  str    
 4   CONTENIDO A ANALIZAR  574761 non-null  str    
 5   INTENSIDAD            574630 non-null  float64
 6   TIPO DE ODIO          12293 non-null   str    
 7   TONO HUMORISTICO      143 non-null     str    
 8   MODIFICADOR           350 non-null     str    
 9   Unnamed: 9            0 non-null       float64
 10  Unnamed: 10           0 non-null       float64
 11  Unnamed: 11           0 non-null       float64
 12  Unnamed: 12           0 non-null       float64
 13  Unnamed: 13           0 non-null       float64
 14  Unnamed: 14           0 non-null       float64
 15  Unnamed: 15

In [6]:
#Elimino los valores nulos en las columnas "Contenido a Analizar", "Tipo de Mensaje" e "intensidad"
data=data.dropna(subset=['CONTENIDO A ANALIZAR',"TIPO DE MENSAJE","INTENSIDAD"]);

#Verificar si existen mas null values
print(data.isnull().sum());

#REviso de nuevo la estructura del Dataframe
print(data.info());



MEDIO                        0
SOPORTE                      0
URL                          0
TIPO DE MENSAJE              0
CONTENIDO A ANALIZAR         0
INTENSIDAD                   0
TIPO DE ODIO            562336
TONO HUMORISTICO        574486
MODIFICADOR             574279
Unnamed: 9              574629
Unnamed: 10             574629
Unnamed: 11             574629
Unnamed: 12             574629
Unnamed: 13             574629
Unnamed: 14             574629
Unnamed: 15             574629
dtype: int64
<class 'pandas.DataFrame'>
Index: 574629 entries, 0 to 575028
Data columns (total 16 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   MEDIO                 574629 non-null  str    
 1   SOPORTE               574629 non-null  str    
 2   URL                   574629 non-null  str    
 3   TIPO DE MENSAJE       574629 non-null  str    
 4   CONTENIDO A ANALIZAR  574629 non-null  str    
 5   INTENSIDAD            57462

In [7]:
#Elimino las columnas Unnamed que solo aportan ruido al dataset
data = data.dropna(how='all', axis=1)

# Verificamos
print(data.info())

<class 'pandas.DataFrame'>
Index: 574629 entries, 0 to 575028
Data columns (total 9 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   MEDIO                 574629 non-null  str    
 1   SOPORTE               574629 non-null  str    
 2   URL                   574629 non-null  str    
 3   TIPO DE MENSAJE       574629 non-null  str    
 4   CONTENIDO A ANALIZAR  574629 non-null  str    
 5   INTENSIDAD            574629 non-null  float64
 6   TIPO DE ODIO          12293 non-null   str    
 7   TONO HUMORISTICO      143 non-null     str    
 8   MODIFICADOR           350 non-null     str    
dtypes: float64(1), str(8)
memory usage: 43.8 MB
None


In [8]:
print(f"Después de eliminar los valores null(NaN) del dataset, el corpus tiene en total {data.shape[0]} registros")

Después de eliminar los valores null(NaN) del dataset, el corpus tiene en total 574629 registros


**Incluye aquí, debajo de la línea, la explicación de tu respuesta**<br><br>
A partir de la carga del dataset **02Dataset_anonimizado.csv**, se procede a guardarlo bajo la variable ***data***, al analizar el dataset, primero reviso que  hayan valores nulos; y efectivamente los hay, entonces procedo a eliminar los registros que contengan esos valores nulos ya que no aportan nada al análisis de la actividad. Igualmente, las columnas Unnamed que todos sus valores son nulos también son eliminadas del dataset. <br>
Luego, para determinar el número de registros post-limpieza,  procedo a utilizar el atributo shape, que si bien me arroja una tupla con el número de registros x número de columnas del dataframe para entender su tamaño, solamente me interesa el número de registros, por lo que selecciono la primera posición de la tupla.
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 2.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántas palabras totales hay en los comentarios del corpus?</span>

In [9]:
#Determino cuantos tipos de mensajes distintos hay en el dataset
data['TIPO DE MENSAJE'].value_counts()

TIPO DE MENSAJE
COMENTARIO         334318
NOTICIA            159413
TITULAR NOTICIA     80898
Name: count, dtype: int64

In [10]:
# Primero, filtro el dataset para que el tipo de mensaje sea "COMENTARIO"
data=data[data["TIPO DE MENSAJE"]=='COMENTARIO'];
#Observo una muestra del nuevo dataset filtrado
data.loc[:, ["TIPO DE MENSAJE", "CONTENIDO A ANALIZAR", "INTENSIDAD"]].head(10)

,TIPO DE MENSAJE,CONTENIDO A ANALIZAR,INTENSIDAD
0,COMENTARIO,el barí§a nunca acaeza ante un segundo b ni an...,3.0
1,COMENTARIO,el real madrid ha puesto punto y final a su an...,0.0
2,COMENTARIO,cristina cifuentes podrí­a haber sido la presi...,3.0
3,COMENTARIO,habrí­a que reabrir el caso. el supremo se ded...,3.0
4,COMENTARIO,me parece un poco exagerado pedir más de tres ...,3.0
5,COMENTARIO,parece que todos los delincuentes niegan las e...,3.0
6,COMENTARIO,"preguntárselo al fracasado, media carrera en 6...",4.0
7,COMENTARIO,tenemos aí±os para ver a esta asociación delic...,3.0
8,COMENTARIO,"una ""fardera"" como presidenta de la comunidad ...",4.0
9,COMENTARIO,uno a uno tienen que ir cayendo frente a la ju...,3.0


In [11]:
#Compruebo que Tipo de mensaje sea solamente "COMENTARIO"
print(data['TIPO DE MENSAJE'].value_counts(),'\n');
print('\n',data.info());

TIPO DE MENSAJE
COMENTARIO    334318
Name: count, dtype: int64 

<class 'pandas.DataFrame'>
Index: 334318 entries, 0 to 575028
Data columns (total 9 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   MEDIO                 334318 non-null  str    
 1   SOPORTE               334318 non-null  str    
 2   URL                   334318 non-null  str    
 3   TIPO DE MENSAJE       334318 non-null  str    
 4   CONTENIDO A ANALIZAR  334318 non-null  str    
 5   INTENSIDAD            334318 non-null  float64
 6   TIPO DE ODIO          10618 non-null   str    
 7   TONO HUMORISTICO      117 non-null     str    
 8   MODIFICADOR           294 non-null     str    
dtypes: float64(1), str(8)
memory usage: 25.5 MB

 None


In [12]:
#Al tener el dataset filtrado solamente por COMENTARIOS, entonces procedo a solo tomar una muestra de 25000 filas de ese dataset(determinado por la variable lines_number) 
#para poder resolver las próximas ejecicios de la actvidad
lines_number = 25000
#data=data.sample(lines_number, random_state=42)
data=data.head(lines_number)
print("Ahora el dataset contiene:", data.shape[0],"registros", "\n" );
print(data.info());

Ahora el dataset contiene: 25000 registros 

<class 'pandas.DataFrame'>
Index: 25000 entries, 0 to 25118
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   MEDIO                 25000 non-null  str    
 1   SOPORTE               25000 non-null  str    
 2   URL                   25000 non-null  str    
 3   TIPO DE MENSAJE       25000 non-null  str    
 4   CONTENIDO A ANALIZAR  25000 non-null  str    
 5   INTENSIDAD            25000 non-null  float64
 6   TIPO DE ODIO          635 non-null    str    
 7   TONO HUMORISTICO      20 non-null     str    
 8   MODIFICADOR           32 non-null     str    
dtypes: float64(1), str(8)
memory usage: 1.9 MB
None


In [13]:
#Muestra de los primeros comentarios
#Se observa errores de caracteres especiales
comentario=data['CONTENIDO A ANALIZAR'].to_list()
print(data['CONTENIDO A ANALIZAR'][20:50].to_list())

['actuar de buena fe es simplemente que a uno le digan "todos lo hacen" para que alguien, con tan pocos fundamentos democraticos, se apunte a lo que sea necesario sin hacerse mas preguntas. las universidades del pp en madrid, capaces de dar masters a sus pí¹pilos a cambio de jugosas subvenciones. y dicen que esto no es una mafia.', 'actuar si que actuó. presuntamente coaccionando a las âÂ\x80Â\x9cprofesorasâÂ\x80Â\x9d. lo de buena fe ya es de traca.', 'cela era de derechas y no daba vergí¼enza ajena. yo dirí\xada más bien que muchos de los polí\xadticos actuales dan vergí¼enza ajena. que de izquierdas también hay alguno que otro...', 'cela.... camilo? y dices que no daba vergí¼enza ajena?', 'como se puede tener tanto morro. toda persona que haya pasado por cualquier universidad espaí±ola sabe que los tí\xadtulos no se dan de oficio, hay que solicitarlos y pagar unas tasas. aunque supongo que como según ella no lo ha pedido tampoco las habrá pagado. qué polí\xadticos tenemos, da hasta

In [14]:
#Uso Regex para limpiar tildes, diéresis, ñ y otros caractéres especiales que fueron cargados pero mal codificados 
#Declarar uso de IA
import re
import unicodedata 
def limpiar_texto(texto):
    # Diccionario manual de reemplazo de caracteres erróneos
    reemplazos = {
        'í±': 'ñ',
        'Ã\x83±': 'ñ',
        'í\xad': 'í',
        'í\xa1': 'á',
        'âÂ\x80Â\x9c': '"',
        'âÂ\x80Â\x9d': '"',
        'í©': 'é',
        'í³': 'ó',
        'íº': 'ú',
        'í§': 'ç',
        'í¼': 'ü' ,
        'ÿ': '¿',
        'Ã\x82¿': '¿',
        'Ã\x82º': 'º',  
        'Ã\x83©':'é',
        'ú': 'º',          
        'Ã\x82º': 'º',
        'á': 'á', 'é': 'é', 'í': 'í', 'ó': 'ó', 'ú': 'ú',
        'Ã\x82Â\xa0': ' ', 'Â\xa0': ' ', '\xa0': ' ',
        'ÿ': '¿', 'Ã\x82¿': '¿', '¿': '¿',
        'á': '¡', 'Ã\x82¡': '¡', '¡': '¡',
        'Ã\x82º': 'º', 'ú': 'º', 'º': 'º',  # Indicador ordinal masculino (nº)
        'Ã\x82ª': 'ª', 'ê': 'ª', 'ª': 'ª',  # Indicador ordinal femenino (1ª)
        'âÂ\x80Â\x9c': '"', 'âÂ\x80Â\x9d': '"', # Comillas tipográficas dobles (“ ”)
        'âÂ\x80Â\x98': "'", 'âÂ\x80Â\x99': "'", # Comillas tipográficas simples (‘ ’)
        'âÂ\x80¦': '...',                      # Puntos suspensivos especiales (…)
        'âÂ\x82¬': '€' 
    }
    for error, correccion in reemplazos.items():
        texto = texto.replace(error, correccion)
    
    # Limpieza de caracteres de control, espacios y puntos consecutivos
    texto = unicodedata.normalize('NFC', texto)
    texto = re.sub(r'[\x00-\x1f\x7f-\x9f\xad]', '', texto)
    texto = re.sub(r'\s+', ' ', texto).strip()
    texto = re.sub(r'\.{2,}', ' ', texto)  
    
    return texto
 #A la misma columna, a cada registro se le aplica la función de limpieza limpiar_texto    
data['CONTENIDO A ANALIZAR'] = [limpiar_texto(t) for t in data['CONTENIDO A ANALIZAR']]
#Se imprime una muestra de la columna para verificar el texto limpio 
print(data['CONTENIDO A ANALIZAR'][20:50].to_list())




['actuar de buena fe es simplemente que a uno le digan "todos lo hacen" para que alguien, con tan pocos fundamentos democraticos, se apunte a lo que sea necesario sin hacerse mas preguntas. las universidades del pp en madrid, capaces de dar masters a sus pí¹pilos a cambio de jugosas subvenciones. y dicen que esto no es una mafia.', 'actuar si que actuó. presuntamente coaccionando a las "profesoras". lo de buena fe ya es de traca.', 'cela era de derechas y no daba vergüenza ajena. yo diría más bien que muchos de los políticos actuales dan vergüenza ajena. que de izquierdas también hay alguno que otro ', 'cela  camilo? y dices que no daba vergüenza ajena?', 'como se puede tener tanto morro. toda persona que haya pasado por cualquier universidad española sabe que los títulos no se dan de oficio, hay que solicitarlos y pagar unas tasas. aunque supongo que como según ella no lo ha pedido tampoco las habrá pagado. qué políticos tenemos, da hasta vergüenza ajena.', 'cualquiera que haya escrit

In [15]:

#Contador de palabras
def contar_palabras_correctas(df):
   #Inicio un contador de palabras en 0
    total_palabras=0 
    #guardo la columna CONTENIDO A ANALIZAR en una variable llamada texto.
    texto=df['CONTENIDO A ANALIZAR'].astype(str).tolist()
    #Un bucle para cada documento preprocesado 
    for doc in nlp.pipe(texto, disable=["ner","parser"]):
        #Creo una lista para determinar si un token es válida(sin tener en cuenta signos de puntuación, sin espacios, sin stop words)
        palabras_valida=[token for token in doc 
                         if not token.is_punct 
                         and not token.is_space 
                         and not token.is_stop 
                         ]
        #Cada vez que tengo una palabra válida, incremento el contador.
        total_palabras=total_palabras + len(palabras_valida)
    return total_palabras


In [16]:
# Aplico la función recién creada al dataset data para determinar el total de palabras.
total_general = contar_palabras_correctas(data)

print(f"Total de palabras en todo el dataset: {total_general}")

Total de palabras en todo el dataset: 981974


***Incluye aquí, debajo de la línea, la explicación de tu respuesta***
<br><br><hr>Lo primero que hago es crear una lista de los comentarios, esto tomando la columna <b>CONTENIDO A ANALIZAR</b>; posteriormente noto que el contenido de la lista tiene caracteres que no fueron bien codificados, o están erróneos, como diéresis, tilde, la ñ, etc...y procedo a limpiarlos, es decir, a remplazarlos por los caracteres correctos. Esto se hizo a través de ReGex. 
<br> A esa misma columna le aplico la función  **limpiar_texto(texto)** y a partir de ahí, con la columna **CONTENID A ANALIZAR** limpia,  es donde se realiza el conteo de palabras. 
<br> Inicializando un contador **total_palabras** y utilizando **nlp.pipe()**, que es un método en spacy útil para el conteo de palabras, defino un bucle en el que para cada token en la lista,  defino que una palabra valida no es ni signo de puntuiación ni es un espacio  ni un stop word y posteriormente, el contador va incrementando  hasta hacer el recorrido por los 25000 registros a analizar. De acuerdo a estas reglas, obtengo un total de ***981974*** palabras .

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 3.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuál el número promedio de palabras en cada comentario?</span>

In [17]:
# El promedio consiste en dividir el total de palabras por el total de registros
promedio_palabras=total_general/lines_number
print(f"{promedio_palabras:.2f}")

39.28


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr> 

Aquí, para responder a esta pregunta, al haber ya obtenido el número total de palabras de todo el corpus, simplemente se trata de dividir ese total de palabras por el número total de registros, en este caso 25000, para un promedio de **39.28** palabras por comentario. 
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 4.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál el número promedio de palabras en los comentarios de cada grupo?</span>

In [18]:
# Función para determinar si el comentario hay odio o no basado en la regla de Intensidad del enunciado
def clasificar_odio(intensidad):
    if intensidad == 0:
        return "No"
    elif intensidad >0:
        return "Sí"
    else : return 'No'

# 2. Se crea la nueva columna 'ODIO' donde se aplica la función creada
data['ODIO'] = data['INTENSIDAD'].apply(clasificar_odio)

# 3. Verificamos los resultados mostrando las primeras filas
print(data[['CONTENIDO A ANALIZAR', 'INTENSIDAD','ODIO']].head(25))


                                 CONTENIDO A ANALIZAR  INTENSIDAD ODIO
0   el barça nunca acaeza ante un segundo b ni ant...         3.0   Sí
1   el real madrid ha puesto punto y final a su an...         0.0   No
2   cristina cifuentes podría haber sido la presid...         3.0   Sí
3   habría que reabrir el caso. el supremo se dedi...         3.0   Sí
4   me parece un poco exagerado pedir más de tres ...         3.0   Sí
5   parece que todos los delincuentes niegan las e...         3.0   Sí
6   preguntárselo al fracasado, media carrera en 6...         4.0   Sí
7   tenemos años para ver a esta asociación delict...         3.0   Sí
8   una "fardera" como presidenta de la comunidad ...         4.0   Sí
9   uno a uno tienen que ir cayendo frente a la ju...         3.0   Sí
10  viéndolo en directo, por si acaso, cualquiera ...         3.0   Sí
11  y ahora esa misma profesora hace esas declarac...         4.0   Sí
12                 ¿eso han dicho? ¿coacción de cifu?         0.0   No
13    

In [19]:
# Conteo para la columna ODIO
conteo_odio = data['ODIO'].value_counts()
print("--- Conteo de la columna ODIO ---")
print(conteo_odio)

# Conteo para la columna INTENSIDAD
conteo_intensidad = data['INTENSIDAD'].value_counts()
print("\n--- Conteo de la columna INTENSIDAD ---")
print(conteo_intensidad)

--- Conteo de la columna ODIO ---
ODIO
No    24365
Sí      635
Name: count, dtype: int64

--- Conteo de la columna INTENSIDAD ---
INTENSIDAD
0.0    24365
4.0      389
1.0       81
3.0       76
2.0       48
5.0       39
6.0        2
Name: count, dtype: int64


In [20]:
# Creo un dataframe filtrado por odio
df_odio = data[data['ODIO'] == 'Sí']
cantidad_odio = len(df_odio)
print(f"Total de mensajes de odio encontrados: {cantidad_odio}")


Total de mensajes de odio encontrados: 635


In [21]:
df_odio.head()

,MEDIO,SOPORTE,URL,TIPO DE MENSAJE,CONTENIDO A ANALIZAR,INTENSIDAD,TIPO DE ODIO,TONO HUMORISTICO,MODIFICADOR,ODIO
0,EL PAÍS,WEB,URL_a4d7efc0,COMENTARIO,el barça nunca acaeza ante un segundo b ni ant...,3.0,Otros,NaN,NaN,Sí
2,EL PAÍS,WEB,URL_54312d9e,COMENTARIO,cristina cifuentes podría haber sido la presid...,3.0,Ideológico,NaN,NaN,Sí
3,EL PAÍS,WEB,URL_54312d9e,COMENTARIO,habría que reabrir el caso. el supremo se dedi...,3.0,Ideológico,NaN,NaN,Sí
4,EL PAÍS,WEB,URL_54312d9e,COMENTARIO,me parece un poco exagerado pedir más de tres ...,3.0,Ideológico,Si,NaN,Sí
5,EL PAÍS,WEB,URL_54312d9e,COMENTARIO,parece que todos los delincuentes niegan las e...,3.0,Ideológico,NaN,NaN,Sí


In [22]:
# Creo un dataframe filtrado sin odio
df_no_odio = data[data['ODIO'] == 'No']
cantidad_no_odio = len(df_no_odio)
print(f"Total de mensajes sin odio encontrados: {cantidad_no_odio}")

Total de mensajes sin odio encontrados: 24365


In [23]:
df_no_odio.head()

,MEDIO,SOPORTE,URL,TIPO DE MENSAJE,CONTENIDO A ANALIZAR,INTENSIDAD,TIPO DE ODIO,TONO HUMORISTICO,MODIFICADOR,ODIO
1,EL PAÍS,WEB,URL_a4d7efc0,COMENTARIO,el real madrid ha puesto punto y final a su an...,0.0,NaN,NaN,NaN,No
12,EL PAÍS,WEB,URL_99b28e63,COMENTARIO,¿eso han dicho? ¿coacción de cifu?,0.0,NaN,NaN,NaN,No
16,EL PAÍS,WEB,URL_99b28e63,COMENTARIO,"¿también sale ayuso en este sumario? caray, me...",0.0,NaN,NaN,NaN,No
18,EL PAÍS,WEB,URL_99b28e63,COMENTARIO,100% pepera. ni una célula fuera de su sitio.,0.0,NaN,NaN,NaN,No
19,EL PAÍS,WEB,URL_99b28e63,COMENTARIO,a mí me la trae floja el resultado. esta señor...,0.0,NaN,NaN,NaN,No


In [24]:
#Funcion para calcular promedio de palabras por cada grupo
def conteo_palabras(datos):
    total_de_palabras=0   
    textos=datos['CONTENIDO A ANALIZAR'].astype(str).tolist()
    
    for doc in nlp.pipe(textos, disable=["parser"]):
        palabras_correctas= [token for token in doc 
                            if not token.is_space 
                            and not token.is_punct
                            and not token.is_stop]
        total_de_palabras+= len(palabras_correctas)
    return total_de_palabras 
   

In [25]:
#A partir dela función creada anteriormente, calculo promedio de palabras en el dataset con odio 
promedio_odio = conteo_palabras(df_odio)/len(df_odio)
#A partir dela función creada anteriormente, calculo promedio de palabras en el dataset sin  odio 
promedio_no_odio = conteo_palabras(df_no_odio)/len(df_no_odio)

# Imprimir resultados
print(f"Promedio de palabras en mensajes con ODIO: {promedio_odio:.2f}")
print(f"Promedio de palabras en mensajes SIN ODIO: {promedio_no_odio:.2f}")

Promedio de palabras en mensajes con ODIO: 8.69
Promedio de palabras en mensajes SIN ODIO: 40.08


**Incluye aquí, debajo de la línea, la explicación de tu respuesta**
<hr> 

Básicamente esto es una combinación de la función creada en la pregunta #2, sin embargo, lo primero que se hace es crear una función para clasificar odio, basado en aplicar la regla de Intensidad y luego aplicar esa función sobre la columna **INTENSIDAD** y ese resultado se guarda en una nueva columna llamada **ODIO**. 
<br> 
Luego, se crea un filtro de odio(toma los registros de ***'CONTENIDO A ANALIZAR'*** donde la columna **ODIO** sea 'Sí') y un filtro de no odio(tomar los registros de 'CONTENIDO A ANALIZAR' donde la columna **ODIO** sea 'No'). Y seguidamente se crea un dataframe que contiene los mensajes de odio y otro dataframe que contiene los mensajes sin odio.
<br>FInalmente se aplica sobre cada uno de esos dataframes la función para el conteo de palabras y se calcula el promedio de palabras en cada dataframe respectivamente. 
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 5.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el número promedio de oraciones en los comentarios de cada grupo?</span>

In [28]:
# Creación de función para calcular el promedio de  oraciones en el dataframe a aplicar.
def promedio_oraciones(df):
    total_comentarios=len(df)
    total_oraciones=0
    datos = df['CONTENIDO A ANALIZAR'].astype(str).tolist()
    for doc in nlp.pipe(datos,disable=['ner']):
         oracion_comentario=len(list(doc.sents))
         total_oraciones += oracion_comentario
    promedio=total_oraciones/total_comentarios
    return promedio       


In [29]:

# Calculamos el promedio para df_odio
promedio_oraciones_odio = promedio_oraciones(df_odio)

# Calculamos el promedio para df_no_odio
promedio_oraciones_no_odio = promedio_oraciones(df_no_odio)

# Imprimimos los resultados con 2 decimales
print("-" * 45)
print(f"Promedio de oraciones en df_odio: {promedio_oraciones_odio:.2f}")
print(f"Promedio de oraciones en df_no_odio: {promedio_oraciones_no_odio:.2f}")
print("-" * 45)

---------------------------------------------
Promedio de oraciones en df_odio: 1.68
Promedio de oraciones en df_no_odio: 2.60
---------------------------------------------


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>

Inicialmente creo una función llamada **promedio_oraciones**, que va a ejecutarla sobre un dataframe(ya sea df_odio o el df_no_odio, que son los dataframes de interés).
<br> 
Posteriormente guardo en la variable **total_comentarios** el total de registros del dataframe que se le aplique a la función y también creo un contador de oraciones **total_oraciones**.
<br> Tomo la columna de texto del dataframe *CONTENIDO A ANALIZAR* y la pongo en una lista. Luego con nlp.pipe preproceso esa columna. Posteriormente doc.sents indica el inicio y fin de una oración, esto finalmente lo pongo en una lista y cuento las oraciones que resulten, dicho resultado se almacena en la variable oracion_comentario. Cada vez que termina el recorrido, entonces incrementa el contador de ***total_oraciones***.
<br> Al terminar de contar las oraciones las divido por el número total de comentarios para obtener el promedio.
<br> Finalmente llamo la función sobre cada uno de los dataframes(df_odio y df_no_odio) para obtener el promedio de oraciones por comentario en cada dataframe. 
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 6.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de comentarios que contienen entidades NER en cada grupo?</span>

In [30]:
# crear funcion de % ner
def pct_ner(df):
    datos=df['CONTENIDO A ANALIZAR'].astype(str).tolist()
    conteo_ner=0
    for doc in nlp.pipe(datos,disable=["parser"]):
        if len(doc.ents)>0:
            conteo_ner+=1
    return (conteo_ner/len(df))*100      
    


In [31]:
# Calculamos los porcentajes
porcentaje_odio = pct_ner(df_odio)
porcentaje_no_odio = pct_ner(df_no_odio)

print(f"Resultados de Análisis NER:")
print(f"- Mensajes con ODIO que contienen entidades: {porcentaje_odio:.2f}%")
print(f"- Mensajes SIN ODIO que contienen entidades: {porcentaje_no_odio:.2f}%")

Resultados de Análisis NER:
- Mensajes con ODIO que contienen entidades: 34.49%
- Mensajes SIN ODIO que contienen entidades: 45.02%


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
<br>
Primero defino una funcion pct_ner, que va a recibir un dataframe df. Creo un contador conteo_ner iniciado en 0 que irá incrementandose cada vez que un mensaje/comentario tenga al menos una entidad. El dataframe que se va a utilizar tomará los datos de la columna CONTENIDO A ANALIZAR. que a su vez los convierto en una lista.
<br>EN la función en sí creo un bucle que para cada documento en la columna CONTENIDO A ANALIZAR detecte si hay alguna entidad (PER,LOC,ORG,MISC) y vaya incrementando el contador y hace esto hasta recorrer la totalidad de los registros del dataframe a utilizar. Luego el promedio se calcula al dividir la cantidad de entidades almancenadas en conteo_ner por el tamaño del dataframe a utilizar. <br>De esta manera, al aplicar esta funcion a los dataframes de df_odio y df_no_odio, se obtiene el porcentaje de entidades por cada registro del df. Los resultados en sí, muestran que para el datafram que contiene odio, de cada 10 mensajes, en 3 hay involucrada una entidad, que podria ser una persona, organización o un lugar. También es posible que este porcentaje se deba a que la mayoría de mensajes con odio no estén dirigidos a una persona u organización específica, sino que se distribuyen de una forma más genérica.
 

<span style="font-size: 16pt; font-weight: bold; color: #0098cd;">Plantea tus propias preguntas</span>

<span><b>Plantea al menos 4 características</b> del texto cuyo análisis permita una caracterización completa del texto. Puedes utilizar recomendaciones proporcionadas por la IA Generativa, si así lo deseas. Para cada una de las características planteadas, obtén valores separados para los grupos ODIO/NO-ODIO.</span>

<span>En la explicación aportada, deberás <b>explicar el significado de la característica planteada</b> así como la importancia de ésta en la caracterización del texto.</span>

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 1.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Dentro del grupo que contiene mensajes de odio y el grupo que contiene mensajes sin odio; ¿A qué personas o a qué organizaciones son dirigidos en su mayoría?</span>

In [32]:

def entidades(df):
   
    # 1. Lista para recolectar las entidades encontradas
    lista_entidades = []
    
    # Etiquetas permitidas
    etiquetas_validas = {'PER',  'ORG'}
    
    # 2. Procesamiento por lotes (nlp.pipe es más rápido para 20k registros)
    # Desactivamos componentes innecesarios para ganar velocidad
    textos = df['CONTENIDO A ANALIZAR'].astype(str)
    
    for doc in nlp.pipe(textos, disable=["parser", "lemmatizer"]):
        for ent in doc.ents:
            # Filtramos por tipo de entidad
            if ent.label_ in etiquetas_validas:
                entidad_texto = ent.text.strip()
                if "USUARIO_OFUSCADO" not in entidad_texto and len(entidad_texto) > 2:
                    lista_entidades.append(entidad_texto.title())                                    
                       
                        
                    
               
                    
    
    # 3. Convertimos la lista a una Serie de Pandas para contar
    # Usamos value_counts y tomamos los primeros 15
    top_15 = pd.Series(lista_entidades).value_counts().head(15)
    
    return top_15

# --- Uso de la función ---
top_entidades_odio = entidades(df_odio)
top_entidades_no_odio= entidades(df_no_odio)
print(top_entidades_odio)

Gentuza               12
Puta                   8
Trump                  7
Psoe                   5
Eta                    2
Franco                 2
San Pedro Zerolo       2
Messi                  1
Cristina Cifuentes     1
Harvard                1
Juan Carlos I          1
Pablo                  1
Pedro                  1
Jesús                  1
Lance Armstrong        1
Name: count, dtype: int64


In [33]:
print(top_entidades_no_odio)

Psoe                  354
Trump                 333
España                128
Oms                   101
Real Madrid           100
Junta De Andalucía     94
Biden                  75
Eta                    75
Salvador Illa          72
Donald Trump           71
Pedro Sánchez          70
Messi                  68
Puig                   63
Tve                    62
Bng                    61
Name: count, dtype: int64


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr>
Entonces, de ese 35% de mensajes que contienen entidades, 
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 2.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Dentro del grupo de mensajes de odio y no odio, Qué tan rico, diverso y único es el vocabulario empleado en cada grupo?</span>

In [34]:
# Creo la funcion de % adjetivos
def type_token_ratio(df):
    palabras_unicas=set()
    total_palabras_validas=0
    textos=df['CONTENIDO A ANALIZAR'].astype(str).tolist()
    for doc in nlp.pipe(textos, disable=['parser', 'ner']):
        palabras_validas=[token for token in doc 
                         if not token.is_punct 
                         and not token.is_space 
                          and not token.is_stop                         
                         ]
        #Cada vez que tengo una palabra válida, incremento el contador.
        total_palabras_validas+= len(palabras_validas)
        #Utilizo lemma para unicamente agregar al conjunto palabras_unicas las palabras diferentes o unicas en cada comentario.
        for token in (palabras_validas):
           palabras_unicas.add(token.lemma_.lower())

    if total_palabras_validas==0: return 0.0
    ttr= (len(palabras_unicas)/total_palabras_validas)*100       
    return ttr          
    
    


In [35]:
 pct_ttr_odio = type_token_ratio(df_odio)
 pct_ttr_no_odio = type_token_ratio(df_no_odio)

 print(f"TTR CON odio: {pct_ttr_odio:.2f}%")
 print(f"TTR SIN odio: {pct_ttr_no_odio:.2f}%")

TTR CON odio: 49.27%
TTR SIN odio: 10.10%


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr>

Dentro de las características que se podrían encontrar, sobretodo en grupos de odio y grupos sin odio, es verificar la diversidad linguistica en cada mensaje escrito para cada grupo. Esto es ***TTR***, Type-Token Ratio, que es una medida de la variabilidad del vocabulario en un texto escrito. Siendo la relación de tokens únicos por tokens totales,el  **TTR** se define así: # Total de Palabras (únicas/Número total de palabras). Donde un alto TTR sugiere un uso diverso del vocabulario y un TTR bajo sugiere uso repetido de terminos y un vocabulario limitado. Ejemplo: "El gato come, el gato duerme, el gato es negro. Quiero a mi gato" tiene 14 palabras en total y 9 types, entonces el TTR=9/14=0.64.
<br>
Básicamente defino una función **type_token_ratio** que va a ser aplicada sobre la columna **CONTENIDO A ANALIZAR**. Inicio un contador de palabras validas en 0 y creo un conjunto de palabras_únicas(esto me va a garantizar que solo palabras diferentes y no repetidas se van a agregar al conjunto). Hago un bucle para cada documento preprocesado(nlp.pipe(...)) y creo una lista de palabras válidas. Luego para cada palabra válida encontrada en la lista, le aplico la funcion token.lemma_  para solamente agregar al conjunto  **palabras_unicas** las palabras diferentes o unicas en cada comentario. Por ejemplo si en la misma oración se encuentra "El perro come. El perro juega", solamente se agregaran al conjunto "El,perro,come,juega"
<br>
Esto muestra que el grupo sin odio muestra una diversidad léxica muy baja, es decir, de cada 40 palabras utilizadas en un mensaje, solo 4 son diferentes en todo el mensaje, las otras 36 son repeticiones.Esto puede sugerir que hay mensajes duplicados.
Por otro lado, para el grupo de mensajes con odio, el valor obtenido fue de 49.27%, es decir,  casi la mitad de las palabras en cada mensaje son diferentes, lo que indica bajo esta medida que hay una diversidad léxica muy alta, lo que puede  sugerir que el discurso de odio no es repetitivo, sino creativo, enfocado a detalle, tiene una capacidad argumentativa variada para atacar.También hay que aclarar que el promedio de palabras por mensaje el el corpus sin odio es alrededor de 40 y con un total de 24365 registros, mientras que en el grupo si odio es de alrededor de 9 palabras por comentario para un total de 635 mensajes. De esta manera, para este corpus en específico resulta complicado interpretar con precisión los valores de TTR arrojados para ambos grupos debido a precisamente la disparidad el el tamaño de las muestras respectivas. En onclusión, mientras el dataset crece, el valor TTR decrece rapidamente.
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 3.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Dentro de los mensajes que contienen odio; ¿Cuáles son los Adjetivos más empleados por quienes emiten estos mensajes?</span>

In [38]:
# Defino función que será aplicada a df_odio y df_no_odio
def adjetivos_odio(df):
    textos=df['CONTENIDO A ANALIZAR']
    #Esta lista guardará la totalidad de adjetivos encontrados
    lista_adjetivos=[]
    #Creo un bucle para cada documento del texto preprocesado
    for doc in nlp.pipe(textos):
        #Para cada token de ese documento, si el pos es adjetivo y no es puntuación ni espacio, lo lemmatiza y lo agrega a la lista
        for token in doc:
            if token.pos_=='ADJ' and not token.is_punct and not token.is_space and not token.is_stop:
                adjetivo=token.lemma_.lower()
                if len(adjetivo)>2:
                    lista_adjetivos.append(adjetivo)
    #La lista la convierto a serie para poder utilizar value_counts y tomar los top 15 basado en ese conteo.
    serie_adjetivos=pd.Series(lista_adjetivos)
    top_15 = serie_adjetivos.value_counts().head(15)
    return top_15

In [39]:
#Procesamiento de Funciones
odio_adj = adjetivos_odio(df_odio)
no_odio_adj=adjetivos_odio(df_no_odio)


In [41]:
print("=== TOP 15 ADJETIVOS MÁS USADOS POR GRUPO DE ODIO ===")
print(odio_adj)
print("=== TOP 15 ADJETIVOS MÁS USADOS POR GRUPO SIN ODIO ===")
print(no_odio_adj)

=== TOP 15 ADJETIVOS MÁS USADOS POR GRUPO DE ODIO ===
español      31
maldito      12
tonto        11
ilegal       11
miserable    10
político      9
público       9
catalán       9
social        8
malo          7
corrupto      7
menudo        7
único         7
asqueroso     7
puto          6
Name: count, dtype: int64
=== TOP 15 ADJETIVOS MÁS USADOS POR GRUPO SIN ODIO ===
social        1878
público       1639
sanitario     1412
alto          1026
necesario      950
municipal      930
importante     869
general        831
local          816
económico      779
español        747
nacional       703
único          648
mayor          636
afectado       635
Name: count, dtype: int64


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr>
<br>
Se elige extraer los adjetivos más utilizados en ambos corpuses, ya que estos son fundamentales en NLP para hacer Análisis de Sentimiento. También, los adjetivos puede funcionar como indicador de alguna carga emocional que se puede hacer más visible en discursos de odio. De ahí que se propone esta característica. 
<br>
Para el código, creo una función que va a actuar sobre el corpus a analizar(odio y no odio), creo una lsita vacía donde voy a agregar todos los adjetivos limpios que encuentre el los documentos a procesar. Luego creo un buble, que para cada documento del texto ya preprocesado(nlp.pipe(texto)...), pongo un condicional sobre si el token es ADJ y no es ni puntuación, ni espacio ni stop word, y si se cumple la condidición, entonces agrego ese palabra a la lista porque es un adjetivo y lo repite hasta el últomo documento. 
Finalmente con la lista, la vuelvo una serie de pandas para con ella poder hacer operaciones de conteo de valores, y a partir de esos valores tomo los 15 adjetivos más usados. Finalmente caluclo la función tanto para el corpus de odio(df_odio) y para el corpus sin odio(df_no_odio).

<br>

En el grupo de odio, los adjetivos claramente son más de tinte despectivo(***maldito,tonto, miserable, malo, asqueroso, p...***), cuyo objetivo es atacar y degradar a una contraparte.

<br>

En contraste con el grupo sin odio, donde los adjetivos son de u corte más objetivo, descriptivo, y, en este caso específico, se enfocan en lo público o política interna***social, público, sanitario, municipal***).

<br>

También hay adjetivos que comparten ambas listas, como "español". Donde en el grupo sin odio, parece un uso meramente descriptivo de un texto, "EL gobierno español", "la sanidad española"
"la economía española"...En contraste con el grupo de odio, en el que denota segregación o exclusión, o superioridad de todo lo español frente a otras identidades o nacionalidades. 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 4.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Dentro del corpus de odio, cuál es el porcentaje de paalabras escirtas en Mayúscula, con signos de exclamacion?</span>

In [45]:
def agresividad(df):
    textos=df['CONTENIDO A ANALIZAR'].astype(str).tolist()
    total_palabras=0
    total_mayusculas=0
    total_tokens=0
    total_exclamaciones=0   
    signos=['!','¡','?','¿']
    

    for doc in nlp.pipe(textos, disable=["parser", "ner","lemmatizer"]):
        for token in doc:
            total_tokens+=1

            for token.text in ['¡','!']:
                total_exclamaciones +=1
            if not token.is_punct and not token.is_space:
                total_palabras+=1

                if token.is_upper and len(token.text)>1:
                    total_mayusculas+=1        

    pct_mayusculas=(total_mayusculas/total_palabras)*100 if total_palabras >0 else 0
    pct_exclamaciones=(total_exclamaciones/total_tokens)*100 if total_tokens >0 else 0
   

    return {
        "% Palabras Mayúsculas": pct_mayusculas,
        "% Exclamaciones": pct_exclamaciones
            }
                    


In [49]:
agresividad_odio=agresividad(df_odio)
agresividad_no_odio=agresividad(df_no_odio)

In [50]:
print(f"GRUPO CON ODIO:")
print(f"- Palabras en Mayúscula: {agresividad_odio['% Palabras Mayúsculas']:.2f}%")
print(f"- Densidad Exclamaciones: {agresividad_odio['% Exclamaciones']:.2f}%")

print("-" * 45)
print(f"GRUPO SIN ODIO:")
print(f"- Palabras en Mayúscula: {agresividad_no_odio['% Palabras Mayúsculas']:.2f}%")
print(f"- Densidad Exclamaciones: {agresividad_no_odio['% Exclamaciones']:.2f}%")


GRUPO CON ODIO:
- Palabras en Mayúscula: 0.70%
- Densidad Exclamaciones: 0.57%
---------------------------------------------
GRUPO SIN ODIO:
- Palabras en Mayúscula: 0.14%
- Densidad Exclamaciones: 0.11%


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr>
<br>
EN NLP, tanto el uso de palabras mayúsculas como de signos de exclamación añaden más información que la palabra sola no puede capturar. La mayúsculas en el mundo digital de las RRSS son el equivalente a gritar. Y Los signos de puntuación pueden denotar alguna alteración, indignación o rabia. Entonces, las mayusculas como los signos pueden ayudar al algoritmo a clasificar mejor el odio, el acoso, a capturar mejor la intensidad emocional en una frase.
<br>
Por lo anterior se propuso función para extraer y calcular la cantidad de palabras escritas en mayúsculsa así como la cantidad de signos de exclamación:
Creo la función que se va a aplicar a los corpuses de interés(odio y no odio). Creo varios contadores, para palabras mayúsculas, signos de exclamación, totalidad de tokens y total de palabras válidas. Luego, para cada documento del texto preprocesado(nlp.pipe(texto...)) cuento el total de tokens y lo almaceno en su respectivo contador.Después, un condicional para extraer los signos de exclamación y al detectarlos, incrementa el respectivo contador de exclamaciones. Luego, para cada token que no sea ni puntuación ni espacio, entonces incrementar el contador de plabras. Y finalmente si el token está es mayúsculas, entonces incrementar el contador de mayúsculas hasta terminar el recorrido por todos los documentos. 
<br>
Finalmente calculo los porcentajes de maypusculas y los porcentajes de exclamaciones y los imprimo.
El reultado arroja, si bien son muy pequeñas las densidades tanto de Mayúsculas com de exclamaciones en el grupo de odio se nota un mayor uso de estos recursos en contraste con el grupo de no odio, Es decir , que el uso de mayusculas y exclamaciones son recursos frecuentemente utilizados por quienes expresan mensajes con odio e intenciones de acentuarlo.

 

<span style="font-size: 16pt; font-weight: bold; color: #0098cd;">Reflexión final</span>

<span>Una de las utilidades de la caracterización de texto es que nos sirve como etapa de <i>feature-extraction</i> (extración de características) de cara a un posterior sistema de clasificación. Es pertinente, por tanto, reflexionar sobre la capacidad discriminatoria de cada una de las características extraídas. </span>

<span> Responde, para ello, a la siguiente pregunta.</span>

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Reflexión final.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Es posible utilizar alguna de las características extraídas en las preguntas anteriores para determinar si un mensaje contiene odio? Justifica tu respuesta con el análisis estadístico que consideres necesario.</span>

In [62]:

def clasificar_odio_dataframe(df):
   
    # lista de lemas tóxicos 
    lemas_toxicos = [
        'puto', 'puta', 'gentuza', 'mierda', 'basura', 'idiota', 'tonto','miserable',
        'imbécil', 'cabrón', 'gilipollas', 'facha', 'ladrón', 'asqueroso','hdlgp','maldito' ]
    
    # Columna a analizar es de tipo texto
    textos = df['CONTENIDO A ANALIZAR'].astype(str).tolist()
    scores = []
    etiquetas = []
    
    # Procesamiento NLP optimizado
    for doc in nlp.pipe(textos, disable=["ner", "parser"]):
        #Creo un contador de score, insultos, palabras mayusculas, signos de exclamaciones y un booleano para determinar si hay o no un pronombre en el documento
        score = 0
        insulto=0
        mayuscula=0
        exclamaciones=0
        hay_pronombre = False

        #Si el documento está vacío, entonces no tiene score y lo etiqueto como No ODio
        if len(doc) == 0:
            scores.append(0)
            etiquetas.append("NO ODIO")
            continue
        #Si el lema está en la lista de lemas tóxicos predefinido, entonces incremento el contador de insultos    
        for token in doc:
            # Léxico Tóxico Lematizado (Suma 30 pts)
            if token.lemma_.lower() in lemas_toxicos:
                insulto+=1
                
            # Detección de ataques directos mediante pronombres (Suma 10 pts)
            if token.pos_ == "PRON":
                hay_pronombre = True
                
                
            # SI la palabra está en maypuscula ,incremento el contador de mayúsculas
            if token.is_upper and len(token.text) > 1:
                mayuscula+=1
                
            # Igual lógica para los signos de exclamación con el contador exclamaciones
            if token.text in ['¡', '!']:
                exclamaciones +=1
                
        # Definición de pesos 
        score +=(insulto * 40)
        if mayuscula >1:
            score+=15
        if exclamaciones >0:
            score +=10
        if insulto>0 and hay_pronombre:
            score+=20
            
        score_final = min(score, 100)
        scores.append(score_final)
        
        # Clasificación final
        if score_final > 50:
            etiquetas.append("ODIO")
        else:
            etiquetas.append("NO ODIO")
            
    # Añadimos los resultados al DataFrame original
    df_resultado = df.copy()
    df_resultado['Hate_Score'] = scores
    df_resultado['Clasificacion_Final'] = etiquetas
    
    return df_resultado




In [63]:
# Llamamos a la función aplicandola al dataframe data
data = clasificar_odio_dataframe(data)


In [64]:
# Vemos un resumen de cuántos clasificó en cada categoría
print("=== RESUMEN DE LA CLASIFICACIÓN ===")
print(data['Clasificacion_Final'].value_counts())
print("\n")

=== RESUMEN DE LA CLASIFICACIÓN ===
Clasificacion_Final
NO ODIO    24210
ODIO         790
Name: count, dtype: int64




In [65]:
# Mostramos los primeros 15 mensajes clasificados como ODIO para validar
pd.set_option('display.max_colwidth', None)
print("=== EJEMPLOS CLASIFICADOS COMO ODIO ===")
mensajes_odio = data[data['Clasificacion_Final'] == 'ODIO']
display(mensajes_odio[['CONTENIDO A ANALIZAR', 'Hate_Score', 'Clasificacion_Final']].head(15))

=== EJEMPLOS CLASIFICADOS COMO ODIO ===


,CONTENIDO A ANALIZAR,Hate_Score,Clasificacion_Final
38,"entonces, no negando la endogamia terrible, me imagino que los no independientes deberían ser los de los contratos basura. no los otros vaya, apaga y vámonos.",60,ODIO
43,pues menuda mierda de igualdad eso si que es discriminación.,60,ODIO
51,"una mierda pal q lo ha escrito, otra para el medio y una a repartir pal q lo crea",60,ODIO
59,pero alguien te escucha? eres tonto chaval?,60,ODIO
77,ya estamos con un chat para i otas! no me creo que el jefe del estado mayor dimita por esto. no somos tontos. no controla a los fascistas dentro del ejercito.,70,ODIO
111,a los fachas no los quiere nadie hasta ellos mismos se niegan.,60,ODIO
124,"que pasa?, no se puede tener una opinión de lo que dice la mayoría, repase las estadísticas de la cantidad de gente que no quiere vacunarse y que ha dicho que hasta que no se vacunen los jefes no lo harían somos hipócritas y criticones por sistema, es la realidad y claro que pienso lo que digo o solo pienso si digo lo que quiere oír la gente?, país democrático, si o dices lo que yo quiero oír o eres, facha, homófobo, tonto lindezas así, pañis, país",100,ODIO
125,"el tribunal supremo debe actuar cuanto antes contra esta pandilla de golpistas indecentes y los compinches prevaricadores sin escrúpulos que tienen colocados en puestos clave para facilitar nuevos golpes y embarrarlo todo del estercolero donde nadan tan a gusto. hay que finiquitar las bulas de estos renegados traidores para sus marrullerías incendiando la convivencia y estafando a la sociedad. p. sanchez y entorno, si esperan tener algún futuro en este país deben mandar a la basura cualquier posibilidad de indulto a estos delincuentes en el trullo y a la fuga. podría traerles la comprensión de un sector de catalanes instalados en la ambigüedad discriminadora supremacista, en el resto de españa, un entierro poco digno y una merecida patada. estos pandillajes involucionistas solo producen hastío y asco. al señor casado y a ciudadanos, menos banderitas, más decencia política y tratar de hacer irrelevantes a cualquier precio a esta pandilla de miserables provincianos fabricantes de odio, en vez de hacerles el juego y regalarles en bandeja, trascendencias que no debieran haber tenido jamás.",60,ODIO
130,"un miserable indigente faccioso menos pero antes tiene que provocar enfrentamientos y desordenes públicos sembrando más odio, del que todavía no se siente suficientemente satisfecho. hemos podido observar en vivo y en directo, de forma cercana y en directo, la explicación veraz de como cualquier fascismo de estercolero con caretas puede llevar a la ruina cualquier democracia.",60,ODIO
131,"un miserable menos pero antes tiene que provocar enfrentamientos y desordenes públicos sembrando más odio, del que todavia no se siente suficientemente satisfecho. hemos podido observar en vivo y en directo de forma cercana y en directo, la explicación veraz de como cualquier fascismo de estercolero con caretas puede llevar a la ruina cualquier democracia. las caretas están de moda y en este caso se han ido al carajo, gracias a dios.",60,ODIO


In [66]:
# Mostramos los primeros 15 mensajes clasificados como NO ODIO para validar
print("=== EJEMPLOS CLASIFICADOS COMO NO ODIO ===")
mensajes_no_odio = data[data['Clasificacion_Final'] == 'NO ODIO']
display(mensajes_no_odio[['CONTENIDO A ANALIZAR', 'Hate_Score', 'Clasificacion_Final']].head(15))

=== EJEMPLOS CLASIFICADOS COMO NO ODIO ===


,CONTENIDO A ANALIZAR,Hate_Score,Clasificacion_Final
0,"el barça nunca acaeza ante un segundo b ni ante un tercera , ya estan los arbitros para impedirlo lo de messi es una autentica vergüenza",0,NO ODIO
1,"el real madrid ha puesto punto y final a su andadura en la copa del rey en el primer escalón. los de zidane han caído ante el alcoyano, de segunda b, a pesar de empezar ganando y jugar con un hombre menos en la prórroga. el técnico francés dispuso un equipo plagado de los menos habituales, con vinicius y mariano en ataque. ninguno de los dos logró crear ocasiones. fue militao el que marcó el gol del madrid, justo antes del descanso. en la segunda parte intentaron cerrar el partido, pero sin el colmillo suficiente y el modesto alcoyano aprovechó un córner para empatar el partido a cinco minutos para el final. el empate sentó como un jarro de agua fría a los blancos, que lo intentaron en el tiempo extra a falta de cinco minutos, el casanova consiguió el gol más importante de su vida, que vale la clasificación para octavos de la copa. el madrid de zidane queda apeado del torneo una vez más, por lo que el francés se quedará sin pelear por el único título que no ha conseguido nunca. así hemos contado el minuto a minuto del partido en directo:",0,NO ODIO
2,"cristina cifuentes podría haber sido la presidenta de madrid que todavía añoran los madrileños, a la vista de lo que es ayuso. podría haber llegado a ser la presidenta del pp, de madrid, por supuesto, pero a nivel estatal quizás también. pero su mala cabeza hundió su futuro político en el fango, de donde es imposible regresar. qué necesidad tenía de demostrar que había sacado un máster, si ni siquiera casado lo tiene. urdió un plan, con amenazas y abuso de poder, para falsificar un documento oficial, penado con cárcel, que no necesitaba para su carrera política. y para rematar el desastre le pillaron hurtando unas cremas faciales que podría haber pagado tranquilamente, pero su cleptomanía oculta y la cámara indiscreta la destrozaron. una vergüenza de libro.",0,NO ODIO
3,habría que reabrir el caso. el supremo se dedica a proteger a los de siempre. hay voluntad para ello?,0,NO ODIO
4,"me parece un poco exagerado pedir más de tres años de prisión a cifuentes por este caso. habría que investigar de quien destapó el caso, al igual que las cremas, seguro que fue de ""fuego amigo"", de alguien que no le interesaba que destapara los casos internos de corrupción en los peperos madrileños como amenazó. al menos esta mujer no se le conoce que se llevara dinero público ni que abasteciera al partido, estorbaba a mucha gente ""amiga"". mucho más grave fue los del título académico de casadete regalado con convalidaciones y su master de tres días de harvard en aravaca. comparada con la actual caótica, no tiene ni punto de comparación, que cuando habla sube el pan a costa de decir sandez tras sandez. mientras, la otra presidenta con título nobiliario, que llenó por lo menos al partido de dinero ilícito, sigue campando por sus anchas.",0,NO ODIO
5,"parece que todos los delincuentes niegan las evidencias y pruebas hasta el ultimo momento. deberia sancionarse a aquellos delincuentes que prorrogan el juicio hasta limites insospechados, con pruebas irrefutables. si la directora en un rasgo de honestidad dice que falsifico y por orden de el ""hagan juego señores, no va mas"" deberia finalizarse. pero para que cifuentes haya delinquido se necesitan cooperadores y complices, de estos hay muchos, empezando por el rector, la asesora etc, etc. en la juan carlos i (vaya una casualidad, dios los cria y ellos se juntan) parece que hay mas irregularidades que en la sede del pp-vox.",0,NO ODIO
6,"preguntárselo al fracasado, media carrera en 6 meses, un máster en aravaca eeuu y otro máster presencial sin ir a clase, este es el nivel de corrupción del pp.",0,NO ODIO
7,"tenemos años para ver a esta asociación delictiva en los banquillos ante los jueces. más agilidad, sres. jueces y que reciban, ni más ni menos l

<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
Si observamos el resultado al aplicar esta nueva función para determinar odio, notamos lo siguiente:
<br>
=== RESUMEN DE LA CLASIFICACIÓN ===<br>
Clasificacion_Final<br>
NO ODIO    24210<br>
ODIO         790<br>

<br>
<br>
Estos resultados son muy similares a los obtenidos en la pregunta pregunta número 4 al hacer el filtrado por mensajes de Odio y no Odio y haber aplicado la regla de intensidad:
<br>

 --- Conteo de la columna ODIO ---<br>
ODIO<br>
No    24365<br>
Sí      635<br>
<br>

Esto da una aproximación casi precisa del sistema de scores planteado en este último ejericio a lo planteado en la pregunta 4 donde se pidió dividir el dataframe en Odio y No Odio basado en una regla previa de la Columna Intensidad. 

Para esta última parte, se planteó inicialmente establecer un sistema de pesos basados en la frecuencia de unos adjetivos condicionados en una lista de insultos previamente establecida, en la presencia o ausencia de palabras en mayúsculas, así como de signos de exclamación y también por la combinación de un pronombre con un adjetivo presente en la lista de insultos. A cada uno de esos items se les da un peso respectivo o score. 
Para la funcion, inicialmente defino la columna de texto a analizar, creo una lista de scores y etiquetas, estas etiquetas al final van a ser rellenadas con "ODIO" o "No ODIO" dependiendo de las condiciones posteriores.
Luego creo un contador de score, insultos, palabras mayusculas, signos de exclamaciones y un booleano para determinar si hay o no un pronombre en el documento a analizar. 
Luego en el bucle: <br>
<br>-Si el documento está vacío, entonces no tiene score y lo etiqueto como No ODio.
<br>-Si el token lematizado está en la lista de lemas tóxicos predefinido, entonces incremento el contador de insultos .
<br>-SI la palabra está en maypuscula ,incremento el contador de mayúsculas
<br>-SI la palabra contiene exclamacoines ,incremento el contador de exclamaciones.
<br>-Defino el sistema de pesos o scores para las ocurrencias de insultos, mayusculas, exclamaciones, y combinaicones de insultos con pronombres.
<br>-Basado en esos pesos, defino si el doc evaluado lleva la etiqueta de ODIO o NO ODIO.
<br>-FInalmente al dataframe creo dos columnas adicionales, Hate_score que va a llevar los scores del cada doc evaluado
<br>-Y la columna de "Clasificación Final" que tomará las etiquetas determinadas para cada doc evaluado.


 

 